### ISSUE TITLE
Support an Infracost SDK

### ISSUE THREAD

**Issue by hugorut** [infracost/infracost#1685]
Users of Infracost can only consume Infracost as a CLI tool. However, many users want to build upon Infracost within internal applications or projects. This issue is a placeholder to discuss and question if Infracost should open an SDK or library that developers can use to augment their projects. Offering such functionality would require refactoring the Infracost internals and is not a trivial change. However, it could offer a much more comprehensive array of applications to use Infracost.

If you are a community member interested in SDK support, please 👍 help us understand the desire for such functionality. Please comment below if you have concrete use cases or requirements for an Infracost SDK.

**Comment by mirusky** [comment]
@hugorut let's discuss a little futher. 

# SDK

I belive that a SDK would be amazing since we could inject the infracost functionalities inside the souce code of any existing product.

# Features

- Custom values list OR extensibility via Hooks / Middlewares 
- Estimate (breakdown)
- Estimate comparison between two configurations

# Explain / Scenarios

- I have some customers that are part of the Azure, AWS and IBM partner programs, so we'd like to pass a custom value list insted of using the global list.
- We'd like to extend some features like while estimating pass a custom value for a resource based on some biz rules (parner program)

**Comment by alikhajeh1** [comment]
@mirusky feel free to join the [August community call](https://github.com/infracost/infracost/issues/1833), it would be useful to better understand the use-cases you have in mind

**Comment by ekristen** [comment]
The easiest path to making this happen is to stop using `internal` directory, replace `internal` with `pkg` then people can start using your code outside of this repo. As of now 99% of the code cannot be imported :(

**Comment by alikhajeh1** [comment]
@ekristen good point - can you possibly share your use-case for an SDK? We're still working on core Infracost concepts - and making big internal changes (e.g. [seeing actual costs too](https://www.infracost.io/docs/infracost_cloud/actual_costs/)), so we're using the Infracost [JSON output](https://www.infracost.io/docs/features/cli_commands/#examples) as our interface to other tools (this is versioned, like the CLI) so we don't break people's CI/CD pipelines. It helps to understand the use-cases as we work on these changes.

**Comment by hugorut** [comment]
Hey @ekristen, exporting all the internals from Infracost is not an option for us. We'd like to maintain a consistent and well-thought-out interface which we can actively support. This is not only simpler for the community/users to adopt but also easier for us to version and provide a stable API.

**Comment by ekristen** [comment]
For me it's less an SDK and getting to reuse some of your code. Like structures for the breakdown or usage file in golang. Or your HCL parser. 

On a personal level I think open source projects should never use internal as it can't be imported externally. It's counter to it being open sourced. 

I'm very appreciative of your tool and the free and open source aspect.

**Comment by glenngillen** [comment]
We're moving to a contribution model where Issues will be exclusively for work that is ready to be worked on. Discussions and feature requests about potential future capabilities will be over in [Discussions](https://github.com/infracost/infracost/discussions). There has been very little community or customer interest in this capability for an extended period, but if it still interests you (or you're now reading this issue because you've been looking for a solution) please [create a new thread in Discussions](https://github.com/infracost/infracost/discussions).
